### DISCUSSION TITLE
Pirate Software Game Jam Retrospective

### DISCUSSION THREAD

**Discussion by McCallisterRomer** [NcStudios/NcEngine#538] [General]
Putting this up as a bucket for thoughts, ideas, bugs, etc. from the game jam. I figured we'd go through in a bit and create issues as needed.

**Comment by McCallisterRomer** [comment]
dev/game-jam branch locked and tag created, so we can rebuild in the future, if needed.

Summary of my commits to dev/game-jam:
- Editor
  - Ability to add [done] `PointLight`/`ConcaveCollider` (already have issues for these)
  - [done]`ParticleEmitter` widget
  - [done] New/Save/Load scene options
  - [done] Wireframe rendering of selected Entity
    - [done] Disabled NC_DEBUG_RENDERING - might not even be used anymore
  - [done] Add Entity can specify layers, flags, etc. (didn't do, but should be done for add child too)
  - [done] `ui::InputU8()` (should just do for all sizes)
  - [done] Disabled asset serialization in scenes (may want an option?)
- Ecs
  - [done]`Ecs::FindEntityByTag()`
  - [done] `Registry::GetEcs()` (might be useful during while finishing the cutover)
  - [done] Entity NoSerialize flag
- Physics
  - [done]Hack to make concave collision ignore static bodies 
- Build/CMake
  - Ability to disable (some) engine install steps
  - Static link CRT (should have a CMake option to select static/dynamic)
- Misc
  - [done] Fixed some `ParticleInfo` defaults
  - [done] Ability to update props in `ParticleEmitter`
  - [done] Fixed broken includes in NcAsset.h
  - [done] Fixed incorrect fwd decl in NcGraphics.h
  - [done] Bump audio buffer size (correct answer might be to make configurable, there may be an issue for this)

**Reply by jaremieromer** [reply]
Summary of my commits to dev/game-jam:



## Vulkan 1.1 Compatibility - All bullets below address: https://github.com/NcStudios/NcEngine/issues/458

 - **Shaders**
	 - Changed version from `#version 460` to `#version 450`
   	- Removed `#GL_EXT_nonuniform_qualifier` and `#GL_ARB_separate_shader_objects`
   	- Hardcoded array length for all descriptors
   	- Made sure all vertex layout entries were present even if unused (eg. `layout (location = 6) in ivec4...`
    - Changed `gl_BaseInstance` to `gl_InstanceIndex`
   
   
 - **Validation Layers**
	 -  Removed copying of `vk_layer_settings.txt`in sample/CMakeLists.txt  so we dump output to the console rather than the file specified in `vk_layer_settings.txt`
	 - Removed `VK_LAYER_LUNARG_api_dump` so we don't dump API info.
 - **C++**
	 - Removed all `vk::DescriptorSetLayoutBindingFlagsCreateInfoExt` structs, so descriptor set no longer uses `vk::DescriptorBindingFlagBitsEXT::ePartiallyBound`
	 - Removed `vk::DescriptorSetLayoutCreateFlagBits::eUpdateAfterBindPool`
	 - Removed device feature requirement for `indexingFeatures.descriptorBindingPartiallyBound && indexingFeatures.runtimeDescriptorArray`. Also removed `vulkan11features.shaderDrawParameters` because it was no longer needed and had been taken from a tutorial prior to that tutorials' update.
	 - Because arrays are now fixed length, modified all image `XShaderResource` classes to `generate_n`, filling the buffer.
	 - Hard-coded very low upper-bound for resource maximums in `ShaderResources.cpp`
	 - Changed Vulkan API version in `IGraphics.cpp` to `VK_API_VERSION_1_1`.
 - **Assets**
	 - Removed `ClickEvents` scene to comply with new lower texture maximums. Ripped out all unused assets (namely, planets, but also those used for `ClickEvents.cpp` and the tree/hatching stuff)
	 - Also removed the unused assets from `Prefabs.cpp`

## Mip Mapping https://github.com/NcStudios/NcEngine/issues/543

 - C++
	 - Parsed mip levels out of image size in `ImmutableImage` and passed into `GpuAllocator`. Propagates through a bunch of `Initializers.cpp` methods.
	 - Added `GenerateMipLevels()` method in `GpuAllocator`

## Font Support - https://github.com/NcStudios/NcEngine/issues/225
 - Added font parsing from file in `Imgui.cpp` 
 
## Cleanup - https://github.com/NcStudios/NcEngine/issues/544
 - Major slashing and simplification of point light/shadow map calculation in `Lighting`, `RenderGraph` and `RenderPass`

**Comment by McCallisterRomer** [comment]
Bugs encountered:

Physics is full of them: Static entities get manipulated sometimes. They're also getting included in collision checks that should be elided earlier on. useGravity and isKinematic aren't always respected? Certain bias/softness values on joints occasionaly failed asserts in DirectXMath.

There are bugs related to transform hierarchies in probably both the editor and serialization. With a child parented to an Entity with a non-identity transform, modifying world space values in the editor is unpredictable (might be good to expose local values too). Loading the same scenario from a prefab seems to perform incorrect multiplication. Maybe we're saving the child's world space values and then double transforming on load? Idk, should check existing tests for coverage of cases like this.

**Reply by jaremieromer** [reply]
We encountered a lot of 'undetectable' validation errors on Phil's laptop that were not present on local machines. These were addressed by adhering largely to the `VK_KHR_portability_subset` specs. 
There are two remaining `VK_KHR_portability_subset` errors:

1.  `max per-stage sampler bindings count (64) exceeds device maxPerStageDescriptorSamplers limit (16)`
2. `pPushConstantRanges[0] offset (0) and size (144) together exceeds this device's maxPushConstantSize of 128. `

**Reply by jaremieromer** [reply]
The point light editor bug was addressed by the `Lighting.cpp` overhaul mentioned above.

**Comment by McCallisterRomer** [comment]
Things that went well:

With recent editor and serialization changes, the sheer amount of code needed is vastly reduced. Its still fairly verbose to create things like particle emitters, but I found myself having to do that on rare occasions, and mostly to make up for things that didn't make the editor in time. Overall, I think the entire game ending up being fewer lines of code than the sample, which is awesome (didn't actually verify this).

Outside of physics and a few editor issues, and known limitations, I hit very few engine issues. I actually can't think of any at the moment. Everything pretty much just worked.

I found it very quick and easy to add the systems and logic necessary, despite having no editor support for these. Cutscene/camera manipulation could be scripted with pretty limited effort. I ended up implementing tree logic a few different ways, all of which were straightforward and non-painful. Collision/Frame/Fixed logic were all super slick. Setting up operations to run on deserialized entities post-scene load went smoothly, although this would probably be more error-prone in larger projects. Even hacking a fade in/out by point light manipulated was pretty easy. I expected these things to be the harder parts but was pleasantly surprised.

**Reply by jaremieromer** [reply]
I have less to add here on overall engine usage because I was largely focused on assets. However, `nc-convert` operated perfectly. It was a breeze to convert assets, and it warned whenever I forgot to do something an asset required (wrong file extension copy-pasta, forgetting to normalize vertex weights, etc).

I also greatly streamlined my asset generation workflow.

**Reply by jaremieromer** [reply]
It was also phenomenal to have a prefab creation editor and a game editor.

**Comment by McCallisterRomer** [comment]
Annoying things:
- Physics. My god.
- `Ecs::Get()` should return a reference. In the game, I wanted the engine to throw on failure 100% of the time. There's no problem checking `Contains()` in the rarer cases.
- Lacking mouse picking + positioning widgets in the editor is a real bummer.
- Workflow for saving/backing up scenes and prefabs was tedious, although this was quickly added for the game and isn't even in vnext yet.
- No ability to copy Entities/Components in the editor slows things down a fair bit.
- As asset counts grow, sifting through dropdowns is a joy sucker.

**Reply by jaremieromer** [reply]
Seconded on the scrolling through dropdowns being a bit pesky, as well as not having copy/paste for entities/components. It would also be GREAT to click and drag in the editor window itself!

**Reply by McCallisterRomer** [reply]
Forgot to mention color over time for particles. They're very limited without it. The ability to toggle billboarding would also have come in handy.
